src/Pate/Verification/Widening.hs

```diff
@@ -171,31 +179,123 @@ widenAlongEdge bundle from d gr to = withSym $ \sym ->
           WideningError msg _ d'' ->
             do let msg' = ("Error during widening: " ++ msg)
                traceBundle bundle msg'
-               postSpec' <- abstractOverVars bundle from to postSpec d''
+               postSpec' <- abstractOverVars scope bundle from to postSpec d''
                case updateDomain gr from to postSpec' of
                  Left gr' ->
                    do traceBundle bundle ("Ran out of gas while widening postconditon! " ++ show from ++ " " ++ show to)
                       return $ recordMiscAnalysisError gr' to (Text.pack msg')
                  Right gr' -> return $ recordMiscAnalysisError gr' to (Text.pack msg')
 
           Widen _ _ d'' -> do
-            postSpec' <- abstractOverVars bundle from to postSpec d''
+            postSpec' <- abstractOverVars scope bundle from to postSpec d''
             case updateDomain gr from to postSpec' of
               Left gr' ->
                 do traceBundle bundle ("Ran out of gas while widening postconditon! " ++ show from ++ " " ++ show to)
                    return gr'
               Right gr' -> return gr'
 
--- TODO: implement this to actually properly perform the scope transformation
+-- | Compute an'PAD.AbstractDomainSpec' from the input 'PAD.AbstractDomain' that is
+-- parameterized over the *output* state of the given 'SimBundle'.
+-- Until now, the widening process has used the same scope for the pre-domain and
+-- post-domain (i.e. both contain free variables corresponding to the initial values
+-- before symbolic execution).
+-- To abstract the computed post-domain from its calling context, we need to rephrase
+-- any symbolic terms it contains so that they only refer to the output state.
+-- Specifically, given a post-domain @dom[pre]@ phrased over the pre-state variables, and
+-- a function @f(pre)@ representing the result of symbolic execution, we want to compute
+-- @dom'[post]@ such that @dom'[post/f(pre)] == dom[pre]@.
 abstractOverVars ::
-  SimBundle sym arch v ->
+  forall sym arch pre.
+  PS.SimScope sym arch pre  ->
+  SimBundle sym arch pre ->
   GraphNode arch {- ^ source node -} ->
   GraphNode arch {- ^ target graph node -} ->
   PAD.AbstractDomainSpec sym arch {- ^ previous post-domain -} ->
-  PAD.AbstractDomain sym arch v {- ^ computed post-domain -} ->
+  PAD.AbstractDomain sym arch pre {- ^ computed post-domain -} ->
```

Comment:
Is this the postdomain with (unsafe) references to variables in the predomain, such that the function fixes up the references to be correct in the returned `AbstractDomainSpec`? I think that interpretation makes sense, but it was a bit confusing to see this argument commented as a post-domain while seeing `pre` in its type.

I think this could just be clarified by tweaking the comment on this argument to say something like `computed post-domain (with unsafe references to pre-domain variables)`

Reply:
Yes that's about right. The resulting 'AbstractDomainSpec' is scope-agnostic (i.e. the body of the spec should only be scoped to the variables from that spec), so we have removed references to the pre-scoped variables in the result.

I added your suggested comment, I'm not sure if this needs to be clarified further though.